packages/require/lib/compiler.ts

```diff
@@ -0,0 +1,94 @@
+import path from "path";
+import TruffleError from "@truffle/error";
+
+import type { CreateOptions, Service } from "ts-node";
+import TruffleConfig from "@truffle/config";
+import { readFileSync, existsSync } from "fs";
+
+import originalRequire from "original-require";
+
+import vm from "vm";
+
+/**
+ * @returns compiler function from user's installation of `ts-node` if script is
+ * a ts or tsx file, otherwise `undefined`.
+ */
+export function compile(conf: TruffleConfig, scriptPath: string): string {
+  const source = readFileSync(scriptPath, { encoding: "utf-8" });
+
+  // we only compile TS files, so just return the source unless we are dealing
+  // with one of those
+  if (![".ts", ".tsx", ".cts"].includes(path.extname(scriptPath))) {
+    return source;
+  }
+
+  const compilationService = _getOrCreateCompilationService(conf, scriptPath);
+
+  return compilationService.compile(source, scriptPath);
+}
+
+export function registerTsNode(
+  conf: TruffleConfig,
+  scriptPath: string,
+  context: Object
+) {
+  // we only compile TS files, so bail unless we are dealing with one of those;
+  if (![".ts", ".tsx", ".cts"].includes(path.extname(scriptPath))) {
+    return;
+  }
+
+  const code = `
+    const tsNode = require("ts-node");
+    const registrationOptions = {
+      cwd: ${JSON.stringify(conf.working_directory.toString())},
+      compilerOptions: {
+        inlineSourceMap: true
+      }
+    };
+    tsNode.register(registrationOptions);
+    `;
+  const script = new vm.Script(code);
+  script.runInContext(context);
+}
+
+let _compilationService: Service | null = null;
+
+function _getOrCreateCompilationService(
+  conf: TruffleConfig,
+  scriptPath: string
+): Service {
+  if (!_compilationService) {
+    try {
+      const createOptions: CreateOptions = {
+        cwd: path.dirname(scriptPath),
+        esm: false,
+        compilerOptions: {
+          inlineSourceMap: true
+        }
+      };
+
+      _compilationService = originalRequire("ts-node").create(createOptions);
+
+      const globalDeclarationsPath = existsSync(
+        path.join(__dirname, "scriptGlobals.ts")
+      )
+        ? path.join(__dirname, "scriptGlobals.ts")
+        : path.join(__dirname, "..", "scriptGlobals.ts");
+      _compilationService!.compile(
+        readFileSync(globalDeclarationsPath, { encoding: "utf-8" }),
+        globalDeclarationsPath
+      );
+    } catch (err) {
+      if (err.code === "MODULE_NOT_FOUND") {
+        throw new TruffleError(`Attempted to execute script with extension
```

Comment:
For debugging it might be nice to have the original error's context as well. Not a blocker for me, but this definitely swallows some information that might be useful. Could `MODULE_NOT_FOUND` have been bubbled up from userland (a user `required` something that isn't installed)?

Reply:
Agreed - will do on both points

Reply:
If only error causes had been added earlier - I might see about adding a polyfill w/i `@truffle/error` to support them, as that'd be much nicer than rolling our own error chaining...

https://v8.dev/features/error-cause
https://node.green/#ES2022-features-Error-cause-property

Reply:
Ah I suppose I don't need to polyfill it - I can just attempt to use it and if it's an early version of node the second argument will be ignored...

Reply:
In order to make this work I need to add error chaining to `@truffle/error`. I think this might be best handled in a separate PR. Should be a quick/easy thing to do, however.